src/printer.rs

```diff
@@ -160,78 +264,94 @@ impl Printer {
-                    self.buffer.print(&body)?;
-                    self.buffer.print("\n\n")?;
+            content_type => {
+                if let Some(body) = request.body().and_then(|b| b.as_bytes()) {
+                    if body.contains(&b'\0') {
+                        self.buffer.print(BINARY_SUPPRESSOR)?;
+                    } else {
+                        self.print_body_text(content_type, &String::from_utf8_lossy(body))?;
+                        self.buffer.print("\n")?;
+                    }
+                    // Breathing room between request and response
+                    self.buffer.print("\n")?;
                 }
             }
-        };
+        }
         Ok(())
     }
 
-    pub async fn print_response_body(&mut self, mut response: Response) -> Result<()> {
-        match get_content_type(&response.headers()) {
-            Some(ContentType::Json) if self.stream => {
-                while let Some(bytes) = response.chunk().await? {
-                    self.print_json(&String::from_utf8_lossy(&bytes))?;
+    pub fn print_response_body(&mut self, mut response: Response) -> anyhow::Result<()> {
+        if !self.buffer.is_terminal() {
+            // No trailing newlines, no decoding, direct streaming
+            self.print_body_stream(get_content_type(&response.headers()), &mut response)?;
+        } else if self.stream {
+            match self.print_body_stream(
+                get_content_type(&response.headers()),
+                &mut decode_stream(&mut response),
+            ) {
+                Ok(_) => {
+                    self.buffer.print("\n")?;
                 }
-            }
-            Some(ContentType::Xml) if self.stream => {
-                while let Some(bytes) = response.chunk().await? {
-                    self.print_xml(&String::from_utf8_lossy(&bytes))?;
-                }
-            }
-            Some(ContentType::Html) if self.stream => {
-                while let Some(bytes) = response.chunk().await? {
-                    self.print_html(&String::from_utf8_lossy(&bytes))?;
-                }
-            }
-            Some(ContentType::Json) => self.print_json(&response.text().await?)?,
-            Some(ContentType::Xml) => self.print_xml(&response.text().await?)?,
-            Some(ContentType::Html) => self.print_html(&response.text().await?)?,
-            _ => {
-                let mut is_first_chunk = true;
-                while let Some(bytes) = response.chunk().await? {
-                    if is_first_chunk
-                        && matches!(self.buffer, Buffer::Stdout | Buffer::Stderr)
-                        && bytes.contains(&b'\0')
-                    {
-                        self.buffer.print(BINARY_SUPPRESSOR)?;
-                        break;
+                Err(err) if err.kind() == io::ErrorKind::InvalidData => {
+                    if self.color {
+                        self.buffer.print("\x1b[0m")?;
```

Comment:
Since Terminal output is the only thing that will ever be suppressed, we can include the reset ASCII code at the beginning of the binary suppressor string.

Reply:
The main reason I do that is that test output becomes *incredibly* confusing if you leave it out. `assert_cmd` just prints the escape code as-is (invisible), so if the test fails because you forgot to put it in you can't see the difference :smile:

But I can also imagine a scenario where someone is using a terminal that doesn't support escape codes and they run `xh` with colors turned off.